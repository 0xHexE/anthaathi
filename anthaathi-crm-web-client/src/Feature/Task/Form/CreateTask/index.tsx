export default function CreateTask() {
  return <form />;
}
