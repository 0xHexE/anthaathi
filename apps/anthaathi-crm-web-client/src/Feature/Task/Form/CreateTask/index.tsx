import * as React from 'react';

export default function CreateTask() {
  return <div>Create task</div>;
}
