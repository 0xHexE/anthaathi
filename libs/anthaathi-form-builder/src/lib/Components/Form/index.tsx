import * as React from 'react';
import {
  ClassType,
  ComponentClass,
  Dispatch,
  FunctionComponent,
  ReactHTML,
  ReactSVG,
  SetStateAction,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { JSONSchema7 } from 'json-schema';
import { apply, RulesLogic } from 'json-logic-js';
import generateInitialJSON from '../../Utils/generate-initial-JSON';
import { FormComponentRegistry } from '../FormComponentRegistryProvider';

export interface Bindings {
  $ref: string;
  $paths: (string | number)[];
}

export type ElementType<T> =
  | UIElement<T>
  | UIElement<T>[]
  | DynamicElement
  | React.ReactNode
  | ModelAccess;

export type TypeComponent =
  | string
  | keyof ReactHTML
  | keyof ReactSVG
  | FunctionComponent
  | ClassType<any, any, any>
  | ComponentClass;

export interface UIElement<T> {
  $import?: string;
  $element: TypeComponent;
  $$kind: 'anthaathi/element';
  scope?: string;
  props?: Record<string, ElementType<T>>;
  binding?: Bindings;
  datasourceBinding?: string | string[];
}

export interface FormProps<T, W = any> {
  $dataSchema: JSONSchema7[];
  $renderSchema: ElementType<T>;
  $dataSources?: Record<string, W>;
}

export interface DynamicElement {
  $$kind: 'anthaathi/dynamic';
  jsonLogic: RulesLogic;
}

export interface ModelAccess {
  $$kind: 'anthaathi/model';
  $path: (string | number)[];
}

export const DataConfigContext = React.createContext<Bindings[]>([]);

export const DataSchemaRegistry = React.createContext<Record<string, any>>({});

export const ConfigProvider = React.createContext<UIElement<any> | null>(null);

export const DataModelRegistry = React.createContext<
  [Record<string, any>, Dispatch<SetStateAction<Record<string, any>>>]
>(null as never);

const useProvideDataSchema = ($dataSchema: JSONSchema7[]) => {
  const initialValues = useMemo(() => {
    const returnObject: Record<string, any> = {};

    $dataSchema.forEach((e) => {
      returnObject[e.$id!] = generateInitialJSON(e);
    });

    return returnObject;
  }, [$dataSchema]);

  return useState<Record<string, any>>(initialValues);
};

export const DataSourceContext = React.createContext<
  [Record<string, any>, Dispatch<SetStateAction<Record<string, any>>>]
>(null as never);

export function Form<T>({
  $dataSchema,
  $renderSchema: $renderSchema_,
  $dataSources = {},
}: FormProps<T>) {
  const state = useState($dataSources || {});
  const [, setDataSources] = state;

  useEffect(() => {
    setDataSources($dataSources);
  }, [$dataSources]);

  const dataSchema = useProvideDataSchema($dataSchema);

  const registry = useContext(FormComponentRegistry);

  const renderComponent = useCallback(
    ($renderSchema: ElementType<T>) => {
      const processedProps: { children?: React.ReactNode } = Object.keys(
        ($renderSchema as UIElement<T>).props || {},
      ).reduce((result, key) => {
        // eslint-disable-next-line no-param-reassign,@typescript-eslint/no-use-before-define
        result[key] = renderElements(
          ($renderSchema as UIElement<T>).props?.[key],
        );

        return result;
      }, {} as Record<string, unknown>);

      const { children, ...props } = processedProps;

      const element = React.createElement(
        ($renderSchema as UIElement<T>).$import
          ? registry[($renderSchema as UIElement<T>).$import as never].find(
              (res) =>
                res.$element === ($renderSchema as UIElement<T>).$element,
            )?.component
          : ($renderSchema as UIElement<T>).$element,
        props,
        children || null,
      );

      return (
        <DataConfigContext.Consumer
          key={($renderSchema as UIElement<T>)?.props?.key as string}
        >
          {(value) => (
            <DataConfigContext.Provider
              value={[
                ...(value || ([] as Bindings[])),
                ($renderSchema as UIElement<T>).binding!,
              ].filter(Boolean)}
            >
              <ConfigProvider.Provider value={$renderSchema as UIElement<T>}>
                {element}
              </ConfigProvider.Provider>
            </DataConfigContext.Provider>
          )}
        </DataConfigContext.Consumer>
      );
    },
    [$renderSchema_, $dataSchema, dataSchema[0]],
  );

  const renderElements = useCallback(
    ($renderSchema: ElementType<T>): React.ReactNode => {
      if (Array.isArray($renderSchema)) {
        return $renderSchema.map((res) => renderElements(res));
      }

      if (($renderSchema as UIElement<T>)?.$$kind === 'anthaathi/element') {
        return renderComponent($renderSchema);
      }

      if (($renderSchema as DynamicElement)?.$$kind === 'anthaathi/dynamic') {
        return apply(
          ($renderSchema as DynamicElement).jsonLogic,
          dataSchema[0],
        );
      }

      if (($renderSchema as ModelAccess)?.$$kind === 'anthaathi/model') {
        return ($renderSchema as ModelAccess).$path.reduce(
          (object, key) => (object || {})[key],
          dataSchema[0],
        ) as any;
      }

      return $renderSchema as React.ReactNode;
    },
    [dataSchema[0], renderComponent],
  );

  const elementToRender = useMemo(
    () => renderElements($renderSchema_) || null,
    [$renderSchema_, dataSchema[0], renderElements],
  );

  return (
    <DataSourceContext.Provider value={state}>
      <DataModelRegistry.Provider value={dataSchema}>
        <DataSchemaRegistry.Provider value={$dataSchema}>
          {elementToRender}
        </DataSchemaRegistry.Provider>
      </DataModelRegistry.Provider>
    </DataSourceContext.Provider>
  );
}

export default Form;
