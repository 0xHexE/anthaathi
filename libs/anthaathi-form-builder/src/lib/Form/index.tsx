import * as React from 'react';
import { useCallback, useMemo } from 'react';
import { JSONSchema7 } from 'json-schema';
import { apply, RulesLogic } from 'json-logic-js';

export interface Bindings {
  $ref: string;
  $paths: (string | number)[];
}

export interface UIElement<T> {
  $import?: string;
  $element: string | React.FunctionComponent;
  $$kind: 'anthaathi/element';
  scope?: string;
  props?: Record<
    string,
    UIElement<T> | UIElement<T>[] | DynamicElement | React.ReactNode
  >;
  binding?: Bindings;
}

export interface FormProps<T> {
  $dataSchema: JSONSchema7;
  $renderSchema: UIElement<T> | UIElement<T>[] | React.ReactNode;
}

export interface DynamicElement {
  $$kind: 'anthaathi/dynamic';
  jsonLogic: RulesLogic;
}

export const DataConfigContext = React.createContext<Bindings[]>([]);

export function Form<T>({
  $dataSchema,
  $renderSchema: $renderSchema_,
}: FormProps<T>) {
  function extracted<TT>(
    $renderSchema:
      | UIElement<T>
      | DynamicElement
      | React.ReactElement<any, string | React.JSXElementConstructor<any>>
      | string
      | number
      | Iterable<React.ReactNode>
      | React.ReactPortal
      | boolean
      | UIElement<TT>[],
  ) {
    /* Processing the props of the element. */
    const processedProps: { children?: React.ReactNode } = Object.keys(
      ($renderSchema as UIElement<TT>).props || {},
    ).reduce((result, key) => {
      // eslint-disable-next-line no-param-reassign
      result[key] = renderElements(($renderSchema as UIElement<T>).props[key]);

      return result;
    }, {});

    const { children, ...props } = processedProps;

    const element = React.createElement(
      ($renderSchema as UIElement<TT>).$element,
      props,
      children || null,
    );

    return (
      <DataConfigContext.Consumer>
        {(value) => (
          <DataConfigContext.Provider
            value={[
              ...(value || []),
              ($renderSchema as UIElement<T>).binding,
            ].filter(Boolean)}
          >
            {element}
          </DataConfigContext.Provider>
        )}
      </DataConfigContext.Consumer>
    );
  }

  const renderElements = useCallback(
    (
      $renderSchema:
        | UIElement<T>
        | UIElement<T>[]
        | DynamicElement
        | React.ReactNode,
    ) => {
      if (Array.isArray($renderSchema)) {
        return $renderSchema.map((res) => renderElements(res));
      }

      if (($renderSchema as UIElement<T>)?.$$kind === 'anthaathi/element') {
        return extracted($renderSchema);
      }

      if (($renderSchema as DynamicElement)?.$$kind === 'anthaathi/dynamic') {
        return apply(($renderSchema as DynamicElement).jsonLogic, {
          test: 'test',
        });
      }

      return $renderSchema;
    },
    [],
  );

  return useMemo(() => renderElements($renderSchema_), [$renderSchema_]);
}

export default Form;
