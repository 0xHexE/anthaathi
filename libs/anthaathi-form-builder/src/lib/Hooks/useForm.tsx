import { useCallback, useContext, useMemo, useState } from 'react';
import { DataConfigContext } from '../Form';

export interface UseFormCallback<T> {
  // eslint-disable-next-line no-unused-vars
  onChange: (_: T) => void;
  onBlur: () => void;
  value?: T;
  error: string[];
}

export function useForm<T>(): UseFormCallback<T> {
  const [value, setValue] = useState<T>();
  const onChange = useCallback((value_: T) => {
    setValue(value_);
  }, []);
  const onBlur = useCallback(() => {}, []);
  const error = useMemo(() => [], []);

  return {
    onChange,
    onBlur,
    value,
    error,
  };
}
