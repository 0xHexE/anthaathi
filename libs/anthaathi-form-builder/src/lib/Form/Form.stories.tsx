import * as React from 'react';
import { Input as InputW, InputProps } from 'baseui/input';
import { FormControl, FormControlProps } from 'baseui/form-control';
import { FC } from 'react';
import { CommonAppProvider } from '@anthaathi/web-lib';
import Form from './index';
import { useForm } from '../Hooks/useForm';

export default {
  title: 'Anthaathi/Form',
  component: Form,
};

function Input() {
  const { onChange, value, onBlur } = useForm<string>();

  return (
    <InputW
      type="text"
      onChange={(e) => onChange((e.target as HTMLInputElement).value)}
      value={value}
      onBlur={() => onBlur()}
    />
  );
}

export function Default() {
  return (
    <CommonAppProvider>
      <Form
        $renderSchema={[
          {
            $element: 'b',
            $$kind: 'anthaathi/element',
            props: {
              key: '12',
              children: {
                $element: 'p',
                $$kind: 'anthaathi/element',
                props: {
                  children: 'secure',
                },
              },
            },
          },
          {
            $element: 'b',
            $$kind: 'anthaathi/element',
            props: {
              key: '122',
              children: {
                $element: 'p',
                $$kind: 'anthaathi/element',
                props: {
                  children: {
                    $$kind: 'anthaathi/element',
                    $element: 'span',
                    props: {
                      children: 'secure1',
                    },
                  },
                },
              },
            },
          },
          {
            $element: InputW as never as FC,
            $$kind: 'anthaathi/element',
            props: {
              value: {
                $$kind: 'anthaathi/model',
                $path: ['https://anthaathi.org/crm/task.json#', 'title'],
              },
              readOnly: true,
            },
            binding: {
              $ref: 'https://anthaathi.org/crm/task.json#',
              $paths: ['title'],
            },
          },
          {
            $element: Input as never,
            $$kind: 'anthaathi/element',
            binding: {
              $ref: 'https://anthaathi.org/crm/task.json#',
              $paths: ['title'],
            },
            props: {
              label: {
                $$kind: 'anthaathi/model',
                $path: ['https://anthaathi.org/crm/task.json#', 'title'],
              },
              placeholder: 'Hello world',
              readOnly: true,
            },
          },
        ]}
        $dataSchema={[
          {
            $id: 'https://anthaathi.org/crm/task.json#',
            type: 'object',
            properties: {
              title: {
                type: 'string',
                maxLength: 255,
                default: 'my name is max',
              },
              description: { type: 'string', maxLength: 5000 },
              requestType: { type: 'string' },
              customer: { type: 'array', items: { type: 'string' } },
              reporter: { type: 'array', items: { type: 'string' } },
              tags: {
                type: 'array',
                items: {
                  type: 'string',
                },
              },
              assigned: {
                type: 'array',
                items: {
                  type: 'string',
                },
              },
              attachment: {
                type: 'string',
              },
              reference: {
                type: 'string',
              },
              priority: {
                type: 'string',
              },
            },
          },
        ]}
      />
    </CommonAppProvider>
  );
}
