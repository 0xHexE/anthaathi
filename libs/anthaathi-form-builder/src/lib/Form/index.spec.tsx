import * as React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Form from './index';
import { useForm } from '../Hooks/useForm';

describe('Form Module', () => {
  function Input(props: React.InputHTMLAttributes<HTMLInputElement>) {
    const { value, onChange, onBlur } = useForm<string>();

    return (
      <input
        type="text"
        value={value}
        onBlur={onBlur}
        onChange={(e) => onChange(e.target.value)}
        {...props}
      />
    );
  }

  it('should render item', () => {
    const snapshot = render(
      <Form
        $dataSchema={[]}
        $renderSchema={{
          $element: 'p',
          $$kind: 'anthaathi/element',
          props: { children: 'test', 'data-id': 'p element' },
        }}
      />,
    );

    expect(snapshot.container).toMatchSnapshot();
  });

  it('should render item array', () => {
    const snapshot = render(
      <Form
        $dataSchema={[]}
        $renderSchema={[
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: { children: 'test', 'data-testid': 'p-element', key: 1 },
          },
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: { children: 'test2', 'data-testid': 'p2-element', key: 2 },
          },
        ]}
      />,
    );

    expect(snapshot.getByTestId('p-element').textContent).toBe('test');
    expect(snapshot.getByTestId('p2-element').textContent).toBe('test2');
  });

  it('should render with two way data binding', () => {
    const snapshot = render(
      <Form
        $dataSchema={[
          {
            $id: 'test',
            type: 'object',
            properties: { title: { type: 'string', default: 'test' } },
          },
        ]}
        $renderSchema={[
          {
            $element: Input,
            $$kind: 'anthaathi/element',
            props: {
              'data-testid': 'input',
              key: 1,
            },
            binding: {
              $ref: 'test',
              $paths: ['title'],
            },
          },
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: {
              children: { $$kind: 'anthaathi/model', $path: ['test', 'title'] },
              'data-testid': 'p2-element',
              key: 2,
            },
          },
        ]}
      />,
    );

    expect((snapshot.getByTestId('input') as HTMLInputElement).value).toBe(
      'test',
    );

    userEvent.type(snapshot.getByTestId('input'), 'This is one of the test');

    expect((snapshot.getByTestId('input') as HTMLInputElement).value).toBe(
      'testThis is one of the test',
    );

    expect(
      (snapshot.getByTestId('p2-element') as HTMLInputElement).textContent,
    ).toBe('testThis is one of the test');

    expect(snapshot.container).toMatchSnapshot();
  });

  it('should render item when it model', () => {
    const snapshot = render(
      <Form
        $dataSchema={[
          {
            $id: 'test',
            type: 'object',
            properties: { title: { type: 'string', default: '' } },
          },
        ]}
        $renderSchema={[
          {
            $element: Input,
            $$kind: 'anthaathi/element',
            props: {
              'data-testid': 'input',
              key: 1,
            },
            binding: {
              $ref: 'test',
              $paths: ['title'],
            },
          },
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: {
              key: 2,
              children: {
                $$kind: 'anthaathi/model',
                $path: ['test', 'title'],
              },
              'data-testid': 'p',
            },
          },
        ]}
      />,
    );

    userEvent.type(snapshot.getByTestId('input'), 'test');
    expect((snapshot.getByTestId('input') as HTMLInputElement).value).toBe(
      'test',
    );
    expect(snapshot.getByTestId('p').textContent).toBe('test');
    expect(snapshot.container).toMatchSnapshot();
  });

  it('should render when item', () => {
    const snapshot = render(
      <Form
        $dataSchema={[
          {
            type: 'object',
            $id: 'test',
            properties: { title: { type: 'string', default: 'test' } },
          },
        ]}
        $renderSchema={[
          {
            $element: Input,
            $$kind: 'anthaathi/element',
            props: {
              'data-testid': 'input',
              key: 2,
            },
            binding: {
              $ref: 'test',
              $paths: ['title'],
            },
          },
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: {
              key: 4,
              'data-testid': 'pElement',
              children: {
                $$kind: 'anthaathi/dynamic',
                jsonLogic: {
                  var: ['test.title'],
                },
              },
            },
          },
          {
            $element: 'p',
            $$kind: 'anthaathi/element',
            props: {
              key: 1,
              'data-testid': 'pElement2',
              children: {
                $$kind: 'anthaathi/dynamic',
                jsonLogic: {
                  cat: [{ var: 'test.title' }, '_TOBE_APPENEDED'],
                },
              },
            },
          },
        ]}
      />,
    );

    expect(snapshot.getByTestId('pElement').textContent).toBe('test');

    userEvent.type(snapshot.getByTestId('input'), 'Something amazing');

    expect(snapshot.container).toMatchSnapshot();
  });
});
