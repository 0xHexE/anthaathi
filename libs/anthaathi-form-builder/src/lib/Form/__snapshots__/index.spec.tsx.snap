// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Module should render item 1`] = `
<div>
  <p
    data-id="p element"
  >
    test
  </p>
</div>
`;

exports[`Form Module should render item when it model 1`] = `
<div>
  <input
    data-testid="input"
    type="text"
    value="test"
  />
  <p
    data-testid="p"
  >
    test
  </p>
</div>
`;

exports[`Form Module should render when item 1`] = `
<div>
  <input
    data-testid="input"
    type="text"
    value="testSomething amazing"
  />
  <p
    data-testid="pElement"
  >
    testSomething amazing
  </p>
  <p
    data-testid="pElement2"
  >
    testSomething amazing_TOBE_APPENEDED
  </p>
</div>
`;

exports[`Form Module should render with two way data binding 1`] = `
<div>
  <input
    data-testid="input"
    type="text"
    value="testThis is one of the test"
  />
  <p
    data-testid="p2-element"
  >
    testThis is one of the test
  </p>
</div>
`;
